virtual_rainforest/core/axes.py

```diff
@@ -0,0 +1,551 @@
+        is assumed to map data onto the cells values defined in the
+        :class:`~virtual_rainforest.core.grid.Grid` configured for the simulation. The
+        coordinate values are assumed to map onto the ``cell_id`` values defined for
+        each cell in the ``Grid``. Because ``cell_id`` values are defined for cells in
+        all grid types, this validator does not require a particular grid geometry.
+    """
+
+    core_axis = "spatial"
+    dim_names = {"cell_id"}
+
+    def can_validate(self, value: DataArray, grid: Grid, **kwargs: Any) -> bool:
+        """Check the validator applies to the inputs.
+
+        Args:
+            value: An input DataArray to check
+            grid: A Grid object giving the spatial configuration of the simulation.
+            kwargs: Other configuration details to be used.
+
+        Returns:
+            A boolean showing if this subclass can be applied to the inputs.
+        """
+        return self.dim_names.issubset(value.dims) and self.dim_names.issubset(
+            value.coords
+        )
+
+    def run_validation(self, value: DataArray, grid: Grid, **kwargs: Any) -> DataArray:
+        """Run validation on the inputs.
+
+        Validation will fail when the ``cell_id`` coordinate values:
+
+        * are not unique, or
+        * do not cover the full set of ``cell_id`` values defined in the configured
+          ``Grid`` object.
+
+        It is permitted for the input DataArray to include a larger set of ``cell_id``
+        coordinates than needed, and the DataArray ``cell_id`` dimension will be reduced
+        to match the ``Grid`` configuration and sorted if necessary.
+
+        Args:
+            value: An input DataArray to check
+            grid: A Grid object giving the spatial configuration of the simulation.
+            kwargs: Other configuration details to be used.
+
+        Raises:
+            ValueError: when ``cell_id`` values are not congruent with the ``Grid``.
+
+        Returns:
+            A DataArray standardised to match the ``cell_id`` values in the ``Grid``
+            object.
+        """
+        da_cell_ids = value["cell_id"].values
+
+        if len(np.unique(da_cell_ids)) != len(da_cell_ids):
+            raise ValueError("The data cell ids contain duplicate values.")
+
+        if not set(grid.cell_id).issubset(da_cell_ids):
+            raise ValueError("The data cell ids are not a superset of grid cell ids.")
+
+        # Now ensure sorting and any subsetting:
+        # https://stackoverflow.com/questions/8251541
+        da_sortorder = np.argsort(da_cell_ids)
+        gridid_pos = np.searchsorted(da_cell_ids[da_sortorder], grid.cell_id)
+        da_indices = da_sortorder[gridid_pos]
+
+        return value.isel(cell_id=da_indices)
+
+
+class Spat_CellId_Dim_Any(AxisValidator):
+    """Validate *cell_id* dimension on the *spatial* core axis.
+
+    Applies to:
+        An input DataArray with a ``cell_id`` dimension is assumed to provide data for
```

Comment:
Think the difference from what the above validator applies could be spelled out a bit more explicitly here, my guess is that this is the one for when the Data array doesn't give cell id values?

Reply:
That's right. You have e.g. 10 values along a `cell_id` dimensions, but those `cell_id` values are just indexed 0-9, rather than explicitly associated with `cell_id` values.